### ISSUE TITLE
try hydra for training configs

### ISSUE THREAD

**Issue by jettjaniak** [delphi-suite/delphi#48]
this is exploratory, shouldn't be done on the actual training script but in isolation as a proof of concept

**Problem:**

imagine:

1. a nested config structure, here represented as a dataclass (in reality this will be model training config)

```python
@dataclass
class Config:
    a: A
    b: B
    c: C
    number: int
    ...

@dataclass
class A:
    x: X
    y: Y
    name: str

@dataclass
class X:
    ...
```

2. a function that does something given an instance of Config (in reality this will be a function that trains a model)
3. script that calls that function

For point 3. to work we need to be able to instantiate Config instance given user input.

**requirements:**
 - use config files but also have an option to override some fields from the command line
 - configs need to be typed, so any configuration mistakes result in loud errors
 - config structure is defined once (e.g. as a dataclass)
 - inheritance
   - if you want to train 10 models that differ only in one config parameter
   - you should be able to define a base config with all parameters
   - and the model-specific config that inherit from the base and override the single parameter
   - i.e. we don't want to copy huge configs just to change a few things

I believe Hydra is right tool for the job. The task is to try it out on a dummy script with dummy config files. You can make a PR and put the configs in static assets dir, I don't expect we will be merging it anyway - if it works we will adapt the training script to use it. 

https://hydra.cc/docs/intro/ (see Structured configs, read the docs, ask chat gpt how to do it, check how other people use Hydra for ML experiments)

**Comment by jaidhyani** [comment]
I think hydra could work for this, but may be overkill and may not be the best fit for the problem. I'm open to being wrong about this, and a working implementation that people like absolutely trumps any theoretical arguments, but here's my reasoning:

1. We want to minimize the number of tools and libraries we're asking developers and users to understand. We should try to only bring in new tools when the problems they solve are more complex than the complexities they introduce. It's not obvious to me that hydra clears this bar for this use case.

2. We're solving for a problem that we don't actually have yet. We can imagine a nested config structure, and experiments that inherit from a common base configuration, but we haven't actually implemented these yet. Hydra is the sort of tool we might look to if we implement those things and *then* find that we want some help wrangling the config logic/making it easier to try new things.

tldr: I think we should be have a basic config solution and a few training runs under our belts before looking into something like this.

**Comment by jettjaniak** [comment]
> We should try to only bring in new tools when the problems they solve are more complex than the complexities they introduce

Agreed. I don't know what complexities Hydra will introduce, that's why I would like someone to try it out.

> We can imagine a nested config structure, and experiments that inherit from a common base configuration, but we haven't actually implemented these yet. Hydra is the sort of tool we might look to if we implement those things and then find that we want some help wrangling the config logic/making it easier to try new things.

I think implementing this ourselves could be more work than using Hydra?

**Comment by jettjaniak** [comment]
Two more arguments for brining Hydra:
1. We have some training runs under our belts, and I have a feel for number of models we're going to train and number of config parameters that go into a training. And I think keeping track of these configs could be a mess.
2. Hydra is widely used in ML experiments. It may be complex, but many will be familiar with it.

**Comment by joshuawe** [comment]
# Config files + command line ✅ 
> use config files but also have an option to override some fields from the command line

Exactly, you can use config files with Hydra. They can be nested etc. And of course you can overwrite individual parameters via the command line. You can also add additional parameters, that are not mentioned in the config this way.
We can use a YAML file or define it in a Python file:
```yaml
model:
  hidden_size: 12
  intermediate_size: 63
  attention_size: 6
  max_position_encoding: 512
  num_attention_heads: 2
  num_hidden_layers: 1
```

Example defining in Python file:
```Python
from dataclasses import dataclass

@dataclass
class ModelConfig:
    hidden_size: int = 12
    intermediate_size: int = 63
    attention_size: int = 6
    max_position_encoding: int = 512
    num_attention_heads: int = 2
    num_hidden_layers: int = 1
```

Example of loading the config file
```Python
import hydra
from omegaconf import OmegaConf
from model_config import ModelConfig

@hydra.main(config_path=".", config_name="config")
def main(cfg) -> None:
    # Convert the loaded configuration to an instance of ModelConfig
    cfg = OmegaConf.to_object(cfg)
    model_cfg = ModelConfig(**cfg.model)
    print("Model Configuration:")
    print(model_cfg)
```



# Parameters need to be typed ✅ 
> configs need to be typed, so any configuration mistakes result in loud errors

If we use a YAML file alone, the data types cannot be specified.
If we also use the Python file, then yes, we should be able to check datatypes as well.

So we could provide a yaml file as well as the following python file:
```Python
from dataclasses import dataclass

@dataclass
class ModelConfig:
    hidden_size: int
    intermediate_size: int
    attention_size: int
    max_position_encoding: int
    num_attention_heads: int
    num_hidden_layers: int
```
~Then the types are set and will be checked by Hydra.~ 
We can use `bearclaw` to make sure these classes and their typehints remain accurate at runtime.

# One definition of config structure ✅ 
> config structure is defined once (e.g. as a dataclass)

Yes, see previous examples.
It is also possible to have config structures in a way, where we include multiple sub options for a parameter e.g. `db` (we want to use different databases for our experiment).

```
├─ conf
│  └─ db
│      ├─ mysql.yaml
│      └─ postgresql.yaml
└── my_app.py
```
Now the configuration has a parameter `db` which can either be of value `mysql` or `postgresql`.
Each has a yaml config file in the `db` folder specifying its parameters.
We can - of course - set a default in our main config file. 
Also, we specify the db parameter in the command line as such:
```
python my_app.py +db=postgresql
```


---
To Do 

> inheritance
>  - if you want to train 10 models that differ only in one config parameter
>  - you should be able to define a base config with all parameters
>  - and the model-specific config that inherit from the base and override the single parameter
>  - i.e. we don't want to copy huge configs just to change a few things

- How to efficiently do sweeps?

**Comment by jaidhyani** [comment]
@joshuawe Check out scripts/run_training.py and the assorted config files in static/config on the `training-script` branch

**Comment by joshuawe** [comment]
> @joshuawe Check out scripts/run_training.py and the assorted config files in static/config on the `training-script` branch

Yes, I skimmed the code regarding the configs and the `GigaConfig` object (great name btw) and all that functionality is basically what Hydra is for, does natively and probably with some extra tools and tricks.

I find it quite simple to use and am creating a basic example setup using Hydra for the Delphi suite, to prove its worth empirically.

**Comment by joshuawe** [comment]
@jettjaniak What is your take away of Hydra? Would you use it in the future?

**Comment by jettjaniak** [comment]
No, I don't think it makes sense to use it for Delphi
